keps/sig-apps/4368-support-managed-by-label-for-batch-jobs/README.md

```diff
@@ -0,0 +1,985 @@
+
+It seems possible, and could be handy in [Story 1](#story-1), but it is also not
+a blocker.
+
+It would also the complicate debuggability of the feature.
+
+We decide to keep the label immutable, at least for [Beta](#beta), we will
+re-evaluate the decision for [GA](#ga).
+
+### Risks and Mitigations
+
+#### Ecosystem fragmentation due to forks
+
+The mechanism to disable the main Job controller opens the door for users to
+substitute it with a fork. This may create more fragmentation in the community
+as users may prefer to use their forked controllers rather than contribute
+upstream.
+
+First, this risk, to some extent, exists even today as admins with access to the
+control plane can disable by passing `--controllers=-job,*` in the manifest for
+`kube-controller-manager` (see more info on the `--controllers` flag
+[here](https://kubernetes.io/docs/reference/command-line-tools-reference/kube-controller-manager/)).
+
+Second, we believe that users who had the need to fork the Job controller
+already introduced dedicated Job CRDs for their needs.
+
+#### Two controllers running at the same time on old version
+
+It is possible that one configures MultiKueue (or another project) with an
+older version of k8s which does not support the label yet. In that case
+two controllers might start running and compete with Job status updates at the
+same time.
+
+To mitigate this risk we warn about it in Kueue documentation, to use this label
+only against newer versions of Kubernetes.
+
+Finally, this risk will fade away with time as the new versions of
+Kubernetes support it.
+
+## Design Details
+
+#### API
+
+```golang
+const (
+  ...
+	// LabelManagedBy is used to indicate the controller or entity that manages
+	// an Job.
+	LabelManagedBy = "batch.kubernetes.io/managed-by"
+)
+```
+
+#### Implementation overview
+
+We skip synchronization of the Jobs with the `managed-by` label, if it has any
+different value than `job-controller.k8s.io`. When the synchronization is skipped,
+the name of the controller managing the Job object is logged.
+
+We leave the particular place at which the synchronization is skipped as
+implementation detail which can be determined during the implementation phase,
+however, two candidate places are:
+1. inside `syncJob` function
+2. inside `enqueueSyncJobInternal` function
+
+Note that, if we skip inside `enqueueSyncJobInternal` we may save on some memory
+needed to needlessly enqueue the Job keys.
+
+There is no validation for the values of the field beyond that of standard
+permitted label values.
+
+#### Label mutability
+
+We keep the label immutable. See also the discussion in
+[Can the label be mutable?](#can-the-label-be-mutable).
+
+### Test Plan
+
+[x] I/we understand the owners of the involved components may require updates to
+existing tests to make this code solid enough prior to committing the changes necessary
+to implement this enhancement.
+
+##### Prerequisite testing updates
+
+##### Unit tests
```

Comment:
> Given that this feature doesn't introduce any new API field, enablement/disablement tests will not provide reasonable value and won't be added.

Shouldn't we test disabling/enabling this feature when there are some jobs with this label set? It seems it might cause incorrect behavior when handling a control from one controller to another. This may be unavoidable, but at least having a test on our end that we are syncing the jobs again could be useful.

Reply:
It makes sense definitely, thank you for pointing out. I have added a note to the unit tests. PTAL.

Reply:
looks ok, thx